language-plutus-core/src/Language/PlutusCore/Instance/Eq/Internal.hs

```diff
@@ -0,0 +1,208 @@
+define a function that records bindings and a function that checks equality of two names in a general
+manner for both the scopes.
+
+In the definitions of equality checks we don't use catch-all patterns when outermost constructors
+don't match, because this way we'd get no warning if someone added a new constructor to a core
+data type and equality checks would silently and erroneously return 'False' for terms containing
+those constructors.
+-}
+
+-- See Note [Modulo alpha].
+-- | Record that two names map to each other.
+withRenamedTwins
+    :: (HasRenaming ren unique, HasUnique name unique, Monad m)
+    => name -> name -> RenameT (Bilateral ren) m c -> RenameT (Bilateral ren) m c
+withRenamedTwins name1 name2 k =
+    withRenamedName (LR name1) (LR name2) $
+    withRenamedName (RL name2) (RL name1) k
+
+-- See Note [Modulo alpha].
+-- | Check equality of two names.
+eqNameM
+    :: (HasRenaming ren unique, HasUnique (name ann) unique, Eq unique)
+    => name ann -> name ann -> EqRename ren
+eqNameM name1 name2 = do
+    mayUniq2' <- lookupNameM $ LR name1
+    mayUniq1' <- lookupNameM $ RL name2
+    let uniq1 = name1 ^. unique
+        uniq2 = name2 ^. unique
+    guard $ case (mayUniq1', mayUniq2') of
+        (Nothing         , Nothing         ) -> uniq1 == uniq2
+        (Just (RL uniq1'), Just (LR uniq2')) -> uniq1 == uniq1' && uniq2 == uniq2'
+        (_               , _               ) -> False
+
+type EqRenameOf ren a = HasUniques a => a -> a -> EqRename ren
+
+-- See Note [Modulo alpha].
+-- | Check equality of two 'Type's.
+eqTypeM :: HasRenaming ren TypeUnique => EqRenameOf ren (Type tyname ann)
+eqTypeM (TyVar _ name1) (TyVar _ name2) =
+    eqNameM name1 name2
+eqTypeM (TyLam _ name1 kind1 ty1) (TyLam _ name2 kind2 ty2) = do
+    guard $ kind1 == kind2
+    withRenamedTwins name1 name2 $ eqTypeM ty1 ty2
+eqTypeM (TyForall _ name1 kind1 ty1) (TyForall _ name2 kind2 ty2) = do
+    guard $ kind1 == kind2
+    withRenamedTwins name1 name2 $ eqTypeM ty1 ty2
+eqTypeM (TyIFix _ pat1 arg1) (TyIFix _ pat2 arg2) = do
+    eqTypeM pat1 pat2
+    eqTypeM arg1 arg2
+eqTypeM (TyApp _ fun1 arg1) (TyApp _ fun2 arg2) = do
+    eqTypeM fun1 fun2
+    eqTypeM arg1 arg2
+eqTypeM (TyFun _ dom1 cod1) (TyFun _ dom2 cod2) = do
+    eqTypeM dom1 dom2
+    eqTypeM cod1 cod2
+eqTypeM (TyBuiltin _ bi1) (TyBuiltin _ bi2) =
+    guard $ bi1 == bi2
+eqTypeM TyVar{}     _ = empty
+eqTypeM TyLam{}     _ = empty
+eqTypeM TyForall{}  _ = empty
+eqTypeM TyIFix{}    _ = empty
+eqTypeM TyApp{}     _ = empty
+eqTypeM TyFun{}     _ = empty
+eqTypeM TyBuiltin{} _ = empty
+
+-- See Note [Modulo alpha].
+-- | Check equality of two 'Term's.
+eqTermM :: EqRenameOf ScopedRenaming (Term tyname name ann)
+eqTermM (LamAbs _ name1 ty1 body1) (LamAbs _ name2 ty2 body2) = do
+    eqTypeM ty1 ty2
+    withRenamedTwins name1 name2 $ eqTermM body1 body2
+eqTermM (TyAbs _ name1 kind1 body1) (TyAbs _ name2 kind2 body2) = do
+    guard $ kind1 == kind2
```

Comment:
I'd probably define `eqKindM` for consistency, and as an obvious place to put a note about the fact that we can use simple equality for kinds.

Reply:
I defined a generic `eqM` working for anything that has an `Eq` instance (because I'm not sure how `kind`s are more special than constants or builtins).